detox/src/Detox.js

```diff
@@ -168,20 +168,25 @@ class Detox {
     this._artifactsManager.registerArtifactPlugins(deviceDriver.declareArtifactPlugins());
 
     await this.device.prepare(params);
-    this._isInitialized = true;
-    this._rejectInit = _.noop;
     return this;
   }
 
-  async _assertNoPendingInit() {
-    if (!this._isInitialized) {
-      this._rejectInit(new DetoxRuntimeError({
-        message: 'Aborted detox.init() execution, and now running detox.cleanup()',
-        hint: 'Most likely, your test runner is tearing down the suite due to the timeout error',
-      }));
+  [_assertNoPendingInit]() {
```

Comment:
I mean, the `Deffer` solution is good but I was hoping for a richer construct that could provide an implementation for this. wdyt?

Reply:
@d4vidi, sorry, but your suggestion is very foggy and blurry, I can't really understand what you mean. Could you elaborate more?

Reply:
Sorry, I'll try to shed more light! 😄 ❤️ 
Actually, `Deferred` _is_ sort of what I meant. However, I was suggesting a class/object that would handle all init-status related concerns (i.e. an init-controller). That would include `assertNoPendingInit` in its API. As it holds all init-related state, it'd be easier to implement:
```
if (state === this.PENDING) { // i.e. not Deferred.PENDING
   throw new DetoxRuntimeError(...);
}
```

and in Detox:
```
this.initController.assertNoPendingInit()
```

That would also dismiss the need for most of the currently public API in `Deferred`, such as `status` and `promise`, I believe. Although, you _can_ keep `Deferred` and have it used by the `"InitController"` since `Deferred` seems rather useful in its own...

Reply:
But anyways, I'm OK with whatever you decided, shouldn't block you from moving fwd.

Reply:
I have investigated the situation, and it seems that extracting some `DetoxInitController` would result in a considerable refactoring worth a separate PR. I think we indeed should pay attention to the bloating of `Detox.js` and start cleaning it up, but just not in this PR.